src/biz/workspace/publish_dup.rs

```diff
@@ -0,0 +1,939 @@
+use shared_entity::dto::workspace_dto;
+use shared_entity::dto::workspace_dto::ViewLayout;
+use sqlx::PgPool;
+use std::collections::HashSet;
+use std::{collections::HashMap, sync::Arc};
+use yrs::updates::encoder::Encode;
+use yrs::{Map, MapRef};
+
+use crate::biz::collab::ops::get_latest_collab_encoded;
+
+#[allow(clippy::too_many_arguments)]
+pub async fn duplicate_published_collab_to_workspace(
+  pg_pool: &PgPool,
+  collab_storage: Arc<CollabAccessControlStorage>,
+  dest_uid: i64,
+  publish_view_id: String,
+  dest_workspace_id: String,
+  dest_view_id: String,
+) -> Result<(), AppError> {
+  let copier = PublishCollabDuplicator::new(
+    pg_pool.clone(),
+    collab_storage.clone(),
+    dest_uid,
+    dest_workspace_id,
+    dest_view_id,
+  );
+  copier.duplicate(&publish_view_id).await?;
+  Ok(())
+}
+
+pub struct PublishCollabDuplicator {
+  /// for fetching and writing folder data
+  /// of dest workspace
+  collab_storage: Arc<CollabAccessControlStorage>,
+  /// A map to store the old view_id that was duplicated and new view_id assigned.
+  /// If value is none, it means the view_id is not published.
+  duplicated_refs: HashMap<String, Option<String>>,
+  /// published_database_id -> view_id
+  duplicated_db_main_view: HashMap<String, String>,
+  /// published_database_view_id -> new_view_id
+  duplicated_db_view: HashMap<String, String>,
+  /// new views to be added to the folder
+  /// view_id -> view
+  views_to_add: HashMap<String, View>,
+  /// A list of database linked views to be added to workspace database
+  workspace_databases: HashMap<String, Vec<String>>,
+  /// time of duplication
+  ts_now: i64,
+  /// for fetching published data
+  /// and writing them to dest workspace
+  pg_pool: PgPool,
+  /// user initiating the duplication
+  duplicator_uid: i64,
+  /// workspace to duplicate into
+  dest_workspace_id: String,
+  /// view of workspace to duplicate into
+  dest_view_id: String,
+}
+
+impl PublishCollabDuplicator {
+  pub fn new(
+    pg_pool: PgPool,
+    collab_storage: Arc<CollabAccessControlStorage>,
+    dest_uid: i64,
+    dest_workspace_id: String,
+    dest_view_id: String,
+  ) -> Self {
+    let ts_now = chrono::Utc::now().timestamp();
+    Self {
+      ts_now,
+      duplicated_refs: HashMap::new(),
+      views_to_add: HashMap::new(),
+      workspace_databases: HashMap::new(),
+      duplicated_db_main_view: HashMap::new(),
+      duplicated_db_view: HashMap::new(),
+
+      pg_pool,
+      collab_storage,
+      duplicator_uid: dest_uid,
+      dest_workspace_id,
+      dest_view_id,
+    }
+  }
+
+  async fn duplicate(mut self, publish_view_id: &str) -> Result<(), AppError> {
+    let mut txn = self.pg_pool.begin().await?;
+
+    // new view after deep copy
+    // this is the root of the document/database duplicated
+    let mut root_view = match self
+      .deep_copy(&mut txn, uuid::Uuid::new_v4().to_string(), publish_view_id)
```

Comment:
It's hard to say how CPU intensive the deep copy is. Ideally we don't want to hold transactions for too long (as we have upper limit of concurrent connections in PgPool). 
1. If possible we could prepare blobs first, then begin transaction and insert them.
2. Alternatively if pt.1 uses too much memory, insert collabs one-by-one each one in separate transaction in their view parent order.

Maybe it would be a good idea to also add metrics to measure time spent on duplicating?

Reply:
I have changed to accumulating all the collabs to insert without holding on to a postgres txn as you mentioned. We already have the metrics to measure the frequency and latency of each http endpoint (including this one), that would suffice for now. when we need more thorough analysis I'll add them to be included to the `/metrics`